src/seq.rs

```diff
@@ -9,61 +9,337 @@
+    {
+        let amount = buf.len();
+        let mut len = 0;
+        while len < amount {
+            if let Some(elem) = self.next() {
+                buf[len] = elem;
+                len += 1;
+            } else {
+                // Iterator exhausted; stop early
+                return len;
+            }
+        }
+
+        // Continue, since the iterator was not exhausted
+        for (i, elem) in self.enumerate() {
+            let k = rng.gen_range(0, i + 1 + amount);
+            if k < amount {
+                buf[k] = elem;
+            }
+        }
+        len
+    }
+
+    /// Collects `amount` values at random from the iterator into a vector.
+    ///
+    /// This is equivalent to `choose_multiple_fill` except for the result type.
+    ///
+    /// Although the elements are selected randomly, the order of elements in
+    /// the buffer is neither stable nor fully random. If random ordering is
+    /// desired, shuffle the result.
+    /// 
+    /// The length of the returned vector equals `amount` unless the iterator
+    /// contains insufficient elements, in which case it equals the number of
+    /// elements available.
+    /// 
+    /// Complexity is `O(n)` where `n` is the length of the iterator.
+    #[cfg(feature = "alloc")]
+    fn choose_multiple<R>(mut self, rng: &mut R, amount: usize) -> Vec<Self::Item>
+        where R: Rng + ?Sized
+    {
+        let mut reservoir = Vec::with_capacity(amount);
+         reservoir.extend(self.by_ref().take(amount));
+
+        // Continue unless the iterator was exhausted
+        //
+        // note: this prevents iterators that "restart" from causing problems.
+        // If the iterator stops once, then so do we.
+        if reservoir.len() == amount {
+            for (i, elem) in self.enumerate() {
+                let k = rng.gen_range(0, i + 1 + amount);
+                if let Some(spot) = reservoir.get_mut(k) {
+                    *spot = elem;
+                }
+            }
+        } else {
+            // Don't hang onto extra memory. There is a corner case where
+            // `amount` was much less than `self.len()`.
+            reservoir.shrink_to_fit();
+        }
+        reservoir
+    }
+}
+
+
+impl<T> SliceExt for [T] {
+    type Item = T;
+
+    fn choose<R>(&self, rng: &mut R) -> Option<&Self::Item>
+        where R: Rng + ?Sized
+    {
+        if self.is_empty() {
+            None
+        } else {
+            Some(&self[rng.gen_range(0, self.len())])
+        }
+    }
+
+    fn choose_mut<R>(&mut self, rng: &mut R) -> Option<&mut Self::Item>
+        where R: Rng + ?Sized
+    {
+        if self.is_empty() {
+            None
+        } else {
+            let len = self.len();
+            Some(&mut self[rng.gen_range(0, len)])
+        }
+    }
+
+    #[cfg(feature = "alloc")]
+    fn choose_multiple<R>(&self, rng: &mut R, amount: usize) -> SliceChooseIter<Self, Self::Item>
+        where R: Rng + ?Sized
+    {
```

Comment:
This one could also be made to return a `Result`, but it's a bit ugly. Thoughts? Doesn't seem to have much overhead, other than the ugly code.
```rust
    fn choose_multiple<R>(&self, rng: &mut R, amount: usize) ->
        Result<SliceChooseIter<Self, Self::Item>, SliceChooseIter<Self, Self::Item>>
        where R: Rng + ?Sized
    {
        if amount <= self.len() {
            Ok(SliceChooseIter {
                slice: self,
                _phantom: Default::default(),
                indices: sample_indices(rng, self.len(), amount).into_iter(),
            })
        } else {
            let mut indices: Vec<usize> = (0..self.len()).collect();
            indices.shuffle(rng);
            Err(SliceChooseIter {
                slice: self,
                _phantom: Default::default(),
                indices: indices.into_iter(),
            })
        }
    }
```

Reply:
The neat-freak in me thinks that maybe we should just go with the old approach.

Reply:
"old" meaning without the `Result`? @pitdicker what do you think (also about using `Result` to indicate not enough elements on the other `choose_multiple` above)?

Reply:
"old" meaning as you have it now, not a `Result`. Same with the other `choose_multiple`.

I want to be clear that I really don't have a strong opinion either way. Other than that I think it makes sense to be consistent.